compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/java/UnsafeCompareAndSwapNode.java

```diff
@@ -98,4 +109,48 @@ public LocationIdentity getLocationIdentity() {
     public void lower(LoweringTool tool) {
         tool.getLowerer().lower(this, tool);
     }
+
+    @Override
+    public void virtualize(VirtualizerTool tool) {
+        ValueNode alias = tool.getAlias(object);
+        if (alias instanceof VirtualInstanceNode && offset.isConstant()) {
+
+            VirtualInstanceNode obj = (VirtualInstanceNode) alias;
+
+            int index = resolveFieldIndex(obj);
+
+            if (index >= 0) {
+
+                ValueNode currentValue = tool.getEntry(obj, index);
+                ValueNode expectedAlias = tool.getAlias(this.expected);
+                ValueNode newValueAlias = tool.getAlias(this.newValue);
+
+                LogicNode equalsNode = CompareNode.createCompareNode(EQ, expectedAlias, currentValue, tool.getConstantReflectionProvider(), NodeView.DEFAULT);
+                if (equalsNode instanceof LogicConstantNode) {
+
+                    boolean equals = ((LogicConstantNode) equalsNode).getValue();
+                    if (equals) {
+                        tool.setVirtualEntry(obj, index, newValueAlias);
+                    }
+                    tool.replaceWith(ConstantNode.forBoolean(equals));
+
+                } else if (!(expectedAlias instanceof VirtualObjectNode) && !(newValueAlias instanceof VirtualObjectNode)) {
+                    ValueNode fieldValue = ConditionalNode.create(equalsNode, newValueAlias, currentValue, NodeView.DEFAULT);
```

Comment:
Did you try
```
Object o = new Object();
AtomicReference<Object> a = new AtomicReference<>(o);
return a.compareAndSet(obj1, obj2);
```
?
If `currentValue` is virtual but expected or new values are not? The created conditional node should lead to issues (virtual only on one side).
You should either disable optimization in this case or look at how `ObjectEqualsNode#virtualize` resolves `==` in various cases (note that you might also be able to resolve the case where both `currentValue` and `expectedAlias` are virtual if you want).

Reply:
Just added a test for it: https://github.com/oracle/graal/pull/636/files#diff-4039dbafeadfd69113c08dcedbdd913cR114.

It takes the branch where both values are non-virtual and virtualizes the CAS. I'm also running this version in a large codebase without problems.

Looking at the `ObjectEquals` virtualization, it'll infer that the the objects are not equal if one of them is virtual and the other is not, so it should be ok.

Reply:
If you return `a.get()` rather than the result of the CAS, you'll see in the graph that it returns a `ConditionalNode` with a `Virtual` node on one side an a `LoadField` on the other, although this test does not fail, we can not generate valid code for that, it would just crash later in the compilation process. See for example this test:
```java
public static Object onlyInitialValueVirtualMatch2() {
    AtomicReference<Object> a = new AtomicReference<>(new Object());
    a.compareAndSet(obj1, obj2);
    return a.get();
}

@Test
public void onlyInitialValueVirtualMatchTest2() {
    testEscapeAnalysis("onlyInitialValueVirtualMatch2", null, true);
    assertTrue(graph.getNodes(LogicCompareAndSwapNode.TYPE).isEmpty());
    test("onlyInitialValueVirtualMatch2");
}
```
This last line with `test` will try to actually compile and use the resulting code which fails because we don't generate code for virtual nodes.

Reply:
Thanks for the example. I've added a condition to check if the current value is not virtual as well.